File: packages/sdk-typescript/src/resolver.ts
/**
 * DNS-over-HTTPS resolver + fetch wrapper. Node.js only.
 *
 * Mirrors the Rust `aex-net` crate and Python SDK's `aex_sdk.resolver`.
 * Uses `tangerine` to perform DoH queries against Cloudflare 1.1.1.1, and
 * `undici` to plug the resolved IP back into the outgoing HTTPS connection
 * via `Agent.connect.lookup`. The TLS SNI + Host header stay set to the
 * original hostname so certificate validation continues to work.
 *
 * ## Browser compatibility
 *
 * This module is Node.js only. In browser contexts, the platform fetch
 * already bypasses OS-level DNS quirks for HTTPS URLs, so DoH is neither
 * necessary nor available. Browser consumers should pass the native
 * `fetch` to {@link SpizeClient} and skip the resolver wiring.
 */
import Tangerine from "tangerine";
import { Agent, fetch as undiciFetch } from "undici";
import { isIP } from "node:net";

/**
 * A DNS resolver that queries Cloudflare DNS-over-HTTPS.
 *
 * Holds no mutable state besides internal tangerine config; safe to share
 * across concurrent callers. Each `resolve` call performs a fresh DoH
 * query — no cache, matching the Rust crate's behaviour.
 */
export class CloudflareDoHResolver {
  private readonly inner: Tangerine;

  constructor() {
    // Tangerine defaults to DoH against Cloudflare; setting servers to
    // the 1.1.1.1 endpoints keeps behaviour explicit and in line with
    // the Rust + Python siblings.
    this.inner = new Tangerine({
      servers: ["1.1.1.1", "1.0.0.1"],
    });
  }

  /** Resolve a hostname to its first A record (IPv4). Throws on failure. */
  async resolve(hostname: string): Promise<string> {
    const raw = (await this.inner.resolve4(hostname)) as
      | string
      | string[]
      | { address: string }[]
      | undefined;
    if (!raw) {
      throw new Error(`no A record returned for ${JSON.stringify(hostname)}`);
    }
    const list = Array.isArray(raw) ? raw : [raw];
    const first = list[0];
    if (first === undefined) {
      throw new Error(`no A record returned for ${JSON.stringify(hostname)}`);
    }
    return typeof first === "string" ? first : first.address;
  }
}

/**
 * Decide whether a given URL needs DoH pre-resolution.
 *
 * Pass-through for: non-HTTPS URLs, IP literals, localhost / single-label
 * hostnames. These don't exercise the system resolver's search-domain path.
 */
export function needsDoh(url: URL): boolean {
  if (url.protocol !== "https:") return false;
  const host = url.hostname;
  if (!host || !host.includes(".")) return false;
  return !isIP(host);
}

/**
 * Build a fetch implementation that pre-resolves HTTPS hostnames via DoH
 * before connecting. Pass-through for HTTP, localhost, and IP literals.
 *
 * Implementation: uses an `undici.Agent` whose `connect.lookup` is driven
 * by the DoH resolver. The outgoing URL's hostname is kept intact (for
 * SNI + cert validation + Host header); only the resolved IP is used at
 * the socket layer.
 */
export function buildDohFetch(
  resolver: CloudflareDoHResolver = new CloudflareDoHResolver(),
): typeof globalThis.fetch {
  const dispatcher = new Agent({
    connect: {
      // Node's lookup contract: (hostname, options, callback)
      lookup: (hostname, _options, callback) => {
        resolver
          .resolve(hostname)
          .then((ip) => callback(null, ip, isIP(ip) || 4))
          .catch((err) => callback(err, "", 0));
      },
    },
  });

  const doFetch = ((input, init) => {
    // Pass-through for requests that don't need DoH.
    const rawUrl =
      typeof input === "string"
        ? input
        : input instanceof URL
          ? input.toString()
          : input.url;
    const url = new URL(rawUrl);
    if (!needsDoh(url)) {
      return globalThis.fetch(input, init);
    }
    return undiciFetch(input as never, {
      ...(init ?? {}),
      dispatcher,
    } as never) as unknown as Promise<Response>;
  }) as typeof globalThis.fetch;

  return doFetch;
}

File: packages/sdk-python/src/aex_sdk/resolver.py
"""DNS-over-HTTPS resolver + httpx transport that bypasses the OS resolver.

Mirrors the Rust `aex-net` crate's ``dns`` + ``http`` modules. See
`docs/protocol-v1.md` §5.3 and the crate docs for the rationale — the short
version is that consumer wifi networks routinely inject a search-domain
suffix into `resolv.conf`, and macOS caches NXDOMAIN for ~60s, either of
which will corrupt lookups of a freshly-minted Cloudflare quick-tunnel
hostname. Using DoH against ``cloudflare-dns.com`` routes around both.
"""

from __future__ import annotations

import ipaddress
from typing import Optional

import dns.message
import dns.query
import dns.rdatatype
import httpx


DEFAULT_DOH_URL = "https://cloudflare-dns.com/dns-query"
DEFAULT_DOH_TIMEOUT = 10.0


class CloudflareDoHResolver:
    """Resolve A records for a hostname via Cloudflare DNS-over-HTTPS.

    The resolver holds no state besides configuration; it is safe to share
    across threads. Each call to :meth:`resolve` performs a fresh DoH query,
    matching the Rust crate's zero-cache behaviour.
    """

    def __init__(
        self,
        doh_url: str = DEFAULT_DOH_URL,
        timeout: float = DEFAULT_DOH_TIMEOUT,
    ) -> None:
        self._doh_url = doh_url
        self._timeout = timeout

    def resolve(self, hostname: str) -> str:
        """Return the first A record for ``hostname`` as a dotted-quad string.

        Raises :class:`RuntimeError` if the DoH query fails or returns no
        A record. Intentionally does not silently fall back to the OS
        resolver — the whole point of this class is to avoid it.
        """
        query = dns.message.make_query(hostname, dns.rdatatype.A)
        # Force HTTP/2 for the DoH transport. dnspython's default tries
        # HTTP/3 first, which raises `NoDOH` when the optional `aioquic`
        # dependency isn't installed. AEX does not require HTTP/3 for
        # DoH — any stack with TLS will do — so we pin H2 to keep the
        # dependency surface small.
        response = dns.query.https(
            query,
            self._doh_url,
            timeout=self._timeout,
            http_version=dns.query.HTTPVersion.H2,
        )
        for rrset in response.answer:
            for rdata in rrset:
                if rdata.rdtype == dns.rdatatype.A:
                    return rdata.address  # type: ignore[no-any-return]
        raise RuntimeError(f"no A record returned for {hostname!r}")


class DoHTransport(httpx.HTTPTransport):
    """httpx transport that pre-resolves hostnames via DoH before connecting.

    Behaviour:

    - For ``https://`` URLs whose host is a DNS name (contains at least one
      dot and is not already an IP literal), the transport asks the resolver
      for the hostname's A record, rewrites the outgoing request's URL to
      point at that IP, sets the ``Host`` header back to the original
      hostname, and sets the httpcore ``sni_hostname`` extension so the TLS
      handshake uses the right SNI and the certificate is verified against
      the hostname.
    - For ``http://``, localhost, short unqualified names, or IP literals,
      the request is passed through unchanged — no DoH lookup, no rewrite.

    This mirrors the ``curl --resolve HOST:443:IP`` pattern the Sprint 1
    demo used via subprocess.
    """

    def __init__(
        self,
        resolver: Optional[CloudflareDoHResolver] = None,
        **transport_kwargs,
    ) -> None:
        super().__init__(**transport_kwargs)
        self._resolver = resolver or CloudflareDoHResolver()

    def handle_request(self, request: httpx.Request) -> httpx.Response:
        if _needs_doh(request.url):
            hostname = request.url.host
            ip = self._resolver.resolve(hostname)
            request.url = request.url.copy_with(host=ip)
            request.headers["host"] = hostname
            request.extensions["sni_hostname"] = hostname.encode()
        return super().handle_request(request)


def _needs_doh(url: httpx.URL) -> bool:
    """True if ``url`` is an HTTPS URL whose host is a public DNS name.

    Pass-through for HTTP (plain), IP literals, and single-label hostnames
    like ``localhost``.
    """
    if url.scheme != "https":
        return False
    host = url.host
    if not host or "." not in host:
        return False
    return not _is_ip_literal(host)


def _is_ip_literal(host: str) -> bool:
    try:
        ipaddress.ip_address(host)
        return True
    except ValueError:
        return False


def build_http_client(
    *,
    resolver: Optional[CloudflareDoHResolver] = None,
    timeout: float = 30.0,
) -> httpx.Client:
    """Return an ``httpx.Client`` that pre-resolves public hostnames via DoH.

    Mirrors the Rust ``aex_net::build_http_client`` factory. Use this for any
    SDK operation that talks to an AEX data plane over a public URL.
    """
    transport = DoHTransport(resolver=resolver or CloudflareDoHResolver())
    return httpx.Client(transport=transport, timeout=timeout)
